8347006: LoadRangeNode floats above array guard in arraycopy intrinsic

TobiHartmann opened the pull request:
C2's arraycopy intrinsic adds guards that check that the source and destination objects are arrays:
https://github.com/openjdk/jdk/blob/afe543414f58a04832d4f07dea88881d64954a0b/src/hotspot/share/opto/library_call.cpp#L5917-L5919

If these guards pass, the array length is loaded:
https://github.com/openjdk/jdk/blob/afe543414f58a04832d4f07dea88881d64954a0b/src/hotspot/share/opto/library_call.cpp#L5930-L5933

But since the `LoadRangeNode` is not pinned, it might float above the array guard:
https://github.com/openjdk/jdk/blob/afe543414f58a04832d4f07dea88881d64954a0b/src/hotspot/share/opto/graphKit.cpp#L1214

If the object is not an array, we will read garbage. That's usually fine because the result will not be used (the array guard will trigger) but with `-XX:+UseCompactObjectHeaders` it can happen that the memory right after the header is not mapped and we crash.

The fix is to add a `CheckCastPPNode` to propagate the information that the operand is an array and prevent the load from floating. 

Thanks to @shipilev for identifying the root cause!

Best regards,
Tobias

rwestrel commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5915,8 +5915,13 @@ bool LibraryCallKit::inline_arraycopy() {
     // (1) src and dest are arrays.
+    // Keep track of the information that src/dest are arrays to prevent below array specific accesses from floating above.
     generate_non_array_guard(load_object_klass(src), slow_region);
+    const Type* tary = TypeAryPtr::make(TypePtr::BotPTR, TypeAry::make(Type::BOTTOM, TypeInt::POS), nullptr, false, Type::OffsetBot);
```
Is this never used else where? Should it a static field in `TypeAryPtr` same as `TypeAryPtr::BYTES` and friends?

TobiHartmann replied:
I wondered as well and no, we don't use this type anywhere else (the closest would be `TypeAryPtr::RANGE`). We only create it when meeting arrays of primitive and non-primitive element type. Do you think this should still go to `TypeAryPtr::*`?

rwestrel replied:
I would add it to `TypeAryPtr` (maybe as `TypeAryPtr::BOTTOM`) . The main benefit I see is that the new code would more readable if it referred to `TypeAryPtr::BOTTOM` rather than the long type creation expression.

TobiHartmann replied:
Sounds good, I'll update the patch accordingly.

rwestrel reviewed (approved):
Looks good to me.

shipilev commented on test/hotspot/jtreg/compiler/arraycopy/TestArrayCopyNoInit.java:
```diff
@@ -23,11 +23,14 @@
  * @run main/othervm -XX:-BackgroundCompilation -XX:-UseOnStackReplacement -XX:TypeProfileLevel=020
  *                   compiler.arraycopy.TestArrayCopyNoInit
+ * @run main/othervm -XX:-BackgroundCompilation -XX:-UseOnStackReplacement -XX:TypeProfileLevel=020
+ *                   -XX:+UnlockExperimentalVMOptions -XX:+UseCompactObjectHeaders -XX:-UseTLAB
```
You have been able to reproduce this with `-UseCompressedClassPointers`, right? If so, I'd suggest we do a run config with `-UseCCP` instead of `+UseCOH`, because this gives us a cleaner way for backports, if we need one later.

TobiHartmann replied:
> You have been able to reproduce this with -UseCompressedClassPointers, right?

No, I was never able to reproduce this with `-XX-UseCompressedClassPointers`.

shipilev replied:
OK, I was confused by this in PR body then:

> I was able to reliably reproduce the issue with compiler/arraycopy/TestArrayCopyNoInit.java and -XX:-UseTLAB -XX:+UnlockExperimentalVMOptions -XX:-UseCompressedClassPointers on Linux AArch64 and verified that the fix solves the problem.

But fine, if it reproduces with +UCOH, let it be there.

TobiHartmann replied:
Ah, that's actually a typo, good catch. Should be `-XX:+UseCompactObjectHeaders`. I'll fix it in the description.

lmesnik replied:
Cant you please add this '@run' as a separate testcase with it's own id. So it is easier to identify and exclude the failures.

shipilev reviewed (commented):
I have a question about the test :)

TobiHartmann commented:
Thanks for the review, Roland!

shipilev commented:
I think bug should target 25 and then we backport it to 24.

TobiHartmann commented:
> I think bug should target 25 and then we backport it to 24.

Since we should fix the issue in JDK 24, the bug should remain targeted to JDK 24. The Skara bot will then take care of updating the fix version to JDK 25 and creating a backport to JDK 24 once we push this into master.

TobiHartmann commented:
I updated the fix according to Roland's suggestions and also added missing stopped checks (otherwise, the Cast will become TOP and below code does not like that).

rwestrel commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5915,8 +5915,16 @@ bool LibraryCallKit::inline_arraycopy() {
     // (1) src and dest are arrays.
+    // Keep track of the information that src/dest are arrays to prevent below array specific accesses from floating above.
     generate_non_array_guard(load_object_klass(src), slow_region);
+    if (!stopped()) {
```
Shouldn't we simply return then?

TobiHartmann replied:
But we need to set up the `slow_region` path, right?

rwestrel replied:
> But we need to set up the `slow_region` path, right?

By that you mean have the `slow_region` feed into the uncommon trap that's only created later. It does feel weird that we know we have reached a dead end and we keep trying to add stuff, but ok then.
The other thing is shouldn't the cast be added in `generate_non_array_guard()`? I see it's used elsewhere (`LibraryCallKit::inline_native_getLength()`): couldn't the same bug occur there?

TobiHartmann replied:
> By that you mean have the slow_region feed into the uncommon trap that's only created later.

Right. It's a bit weird but probably still the best solution in terms of complexity. The trap will lead to recompilation and then the `too_many_traps` check will triggers.

> The other thing is shouldn't the cast be added in generate_non_array_guard()? I see it's used elsewhere (LibraryCallKit::inline_native_getLength()): couldn't the same bug occur there?

Right, good catch. I think the use in `LibraryCallKit::inline_native_getLength` has the same problem. We can't easily put the cast into `generate_non_array_guard` though because it operates on the Klass and not the object. The other `generate*array*guard` methods potentially have the same issue but current uses look good. I guess it's best to fix the `LibraryCallKit::inline_native_getLength` as well, i.e., make it the caller's responsibility. What do you think?

TobiHartmann replied:
Hmm, maybe `inline_getObjectSize` is affected as well:

https://github.com/openjdk/jdk/blob/afe543414f58a04832d4f07dea88881d64954a0b/src/hotspot/share/opto/library_call.cpp#L8535-L8543

rwestrel replied:
> I guess it's best to fix the `LibraryCallKit::inline_native_getLength` as well, i.e., make it the caller's responsibility to add a cast. What do you think?

Maybe the methods need to take an extra parameter (the object to cast)?
Having the cast in the method would lead to less code duplication and a lower risk of forgetting the cast when new calls of the method are added so that's what I would go with unless it's really a pain.

merykitty commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5915,8 +5915,16 @@ bool LibraryCallKit::inline_arraycopy() {
+    // Keep track of the information that src/dest are arrays to prevent below array specific accesses from floating above.
     generate_non_array_guard(load_object_klass(src), slow_region);
+    if (!stopped()) {
+      src = _gvn.transform(new CheckCastPPNode(control(), src, TypeAryPtr::BOTTOM));
```
Why is this a `CheckCastPP` and not a `CastPP`? My understanding is that a `CheckCastPP` is used when we force changing the type of a node (e.g a raw pointer of `Allocate` into a typed pointer), so we do not join the type of the input with that of the output.

TobiHartmann replied:
Good point, I changed that.

merykitty commented on src/hotspot/share/opto/type.hpp:
```diff
@@ -1473,6 +1473,7 @@ class TypeAryPtr : public TypeOopPtr {
   virtual const TypeKlassPtr* as_klass_type(bool try_for_exact = false) const;
 
   // Convenience common pre-built types.
+  static const TypeAryPtr* BOTTOM;
```
While you are here it may be better to change the other constant to `TypeAryPtr*` instead of `TypeAryPtr *`

TobiHartmann replied:
Right, done.

TobiHartmann commented:
Roland, Quan Anh, thanks for the reviews! I pushed a new version that should address all comments.

TobiHartmann commented:
Thanks again for the review, @merykitty!

TobiHartmann commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -4299,7 +4299,14 @@ Node* LibraryCallKit::generate_array_guard_common(Node* kls, RegionNode* region,
+  if (obj != nullptr && is_array_ctrl != nullptr && is_array_ctrl != top()) {
+    // Keep track of the fact that 'obj' is an array to prevent
+    // array specific accesses from floating above the guard.
+    *obj = _gvn.transform(new CastPPNode(is_array_ctrl, *obj, TypeAryPtr::BOTTOM));
```
I thought about this as well but I don't think it's necessary because:
- No cast is needed if we know the type already
- We don't emit a guard and only one branch remains, so there is no risk of the array specific access floating above

So I went with the simplest changes for now, also since we need to backport this change (it got already much more complicated than I was aiming for).

vnkozlov reviewed (approved):
Good.

TobiHartmann commented:
Thanks for the review, Vladimir! @rwestrel are you okay with the changes as well?

TobiHartmann commented:
/integrate

TobiHartmann commented:
/backport :jdk24
